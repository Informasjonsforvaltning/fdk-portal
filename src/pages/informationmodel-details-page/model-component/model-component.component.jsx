import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import _ from 'lodash';
import { Collapse } from 'reactstrap';
import { withStateHandlers } from 'recompose';
import Scroll from 'react-scroll';

import {
  getOwnProperties,
  getParentTypeRef,
  getTypeFromTypeRef
} from '../../../lib/informationmodel/informationmodel';
import { ModelComponentProperty } from './model-component-property/model-component-property.component';
import './model-component.scss';

const ScrollLink = Scroll.Link;

const renderProperties = properties => {
  if (!properties) {
    return null;
  }

  return Object.keys(properties).map(key => {
    const { type, typeRef, oneOfTypeRefs } = properties[key];

    return (
      <ModelComponentProperty
        key={key}
        name={key}
        type={type}
        typeRef={typeRef}
        oneOfTypeRefs={oneOfTypeRefs}
      />
    );
  });
};

const renderName = (collapseIconClass, onToggle, name, propertiesIsEmpty) => {
  if (propertiesIsEmpty) {
    return (
      <div
        style={{ marginLeft: '16px' }}
        className="d-flex align-items-center w-50"
        name={name}
      >
        <span className="d-flex w-50 fdk-text-extra-strong">{name}</span>
      </div>
    );
  }
  return (
    <button
      type="button"
      style={{ marginRight: '16px' }}
      className="bg-transparent border-0 p-0 d-flex align-items-center w-50 node"
      onClick={onToggle}
      name={name}
    >
      <div className="d-flex align-items-center">
        <i className={collapseIconClass} />
        <span className="d-flex w-50 fdk-text-extra-strong">{name}</span>
      </div>
    </button>
  );
};

const renderType = (type, nativeType) => {
  if (type) {
    return (
      <div className="d-flex w-50">
        <ScrollLink
          key={type}
          to={type}
          spy
          smooth
          isDynamic
          offset={-10}
          duration={1500}
          tabIndex="0"
          className="node"
        >
          <span>{type}</span>
        </ScrollLink>
      </div>
    );
  }
  if (!type && nativeType) {
    return <div className="d-flex w-50">{nativeType}</div>;
  }
  return null;
};

export const ModelComponentPure = props => {
  const { name, definitions, collapse, onToggle } = props;

  if (!definitions) {
    return null;
  }

  const definition = definitions[name];
  const type = getTypeFromTypeRef(getParentTypeRef(definition));
  const nativeType = getTypeFromTypeRef(definition.type);
  const properties = getOwnProperties(definition);
  const propertiesIsEmpty = _.isEmpty(getOwnProperties(definition), true);

  const collapseIconClass = cx(
    'fa',
    'font-weight-bold',
    'mr-2',
    'fdk-color-link',
    {
      'fa-angle-down': !collapse,
      'fa-angle-up': collapse
    }
  );

  return (
    <>
      <div className="d-flex">
        {renderName(collapseIconClass, onToggle, name, propertiesIsEmpty)}

        {renderType(type, nativeType)}
      </div>
      <Collapse className="mt-2" isOpen={collapse}>
        {renderProperties(properties)}
      </Collapse>
    </>
  );
};

ModelComponentPure.defaultProps = {
  name: null,
  definitions: null
};

ModelComponentPure.propTypes = {
  name: PropTypes.string,
  definitions: PropTypes.object
};

const enhance = withStateHandlers(
  ({ initialCollapse = false }) => ({
    collapse: initialCollapse
  }),
  {
    onToggle: ({ collapse }) => () => ({ collapse: !collapse })
  }
);

export const ModelComponentWithState = enhance(ModelComponentPure);
export const ModelComponent = ModelComponentWithState;
