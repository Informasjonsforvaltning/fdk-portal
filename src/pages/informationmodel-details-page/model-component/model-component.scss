@import '../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.schema-collapse--item {
  margin-bottom: -1px;
  background-color: $color-white;
  border-top: 1px solid $color-neutral-light;
  padding: 1em 0;
  flex-wrap: wrap !important;
  &:first-child {
    border-top: 0;
  }
  &:last-child {
    border-bottom: 0;
  }
}
