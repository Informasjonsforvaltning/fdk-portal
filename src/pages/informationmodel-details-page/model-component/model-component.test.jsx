import React from 'react';
import { shallow } from 'enzyme';
import { ModelComponentPure } from './model-component.component';
import informationModelApiResponse from '../../../mock/informationmodel.response.json';

const schema = JSON.parse(informationModelApiResponse.schema);

test('should render ModelComponentPure correctly with no props', () => {
  const wrapperWithNoProps = shallow(<ModelComponentPure />);
  expect(wrapperWithNoProps).toMatchSnapshot();
});

test('should render ModelComponentPure with props', () => {
  const defaultProps = {
    name: 'Enhet',
    definitions: schema.definitions
  };
  const wrapper = shallow(<ModelComponentPure {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
