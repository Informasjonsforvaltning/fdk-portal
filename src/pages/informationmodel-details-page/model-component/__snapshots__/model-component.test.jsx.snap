// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ModelComponentPure correctly with no props 1`] = `""`;

exports[`should render ModelComponentPure with props 1`] = `
<Fragment>
  <div
    className="d-flex"
  >
    <button
      className="bg-transparent border-0 p-0 d-flex align-items-center w-50 node"
      name="Enhet"
      style={
        Object {
          "marginRight": "16px",
        }
      }
      type="button"
    >
      <div
        className="d-flex align-items-center"
      >
        <i
          className="fa font-weight-bold mr-2 fdk-color-link fa-angle-down"
        />
        <span
          className="d-flex w-50 fdk-text-extra-strong"
        >
          Enhet
        </span>
      </div>
    </button>
  </div>
  <Collapse
    appear={false}
    className="mt-2"
    enter={true}
    exit={true}
    in={false}
    isOpen={false}
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    tag="div"
    timeout={350}
    unmountOnExit={false}
  >
    <ModelComponentProperty
      key="organisasjonsnummer"
      name="organisasjonsnummer"
      oneOfTypeRefs={null}
      type="string"
      typeRef={null}
    />
    <ModelComponentProperty
      key="navn"
      name="navn"
      oneOfTypeRefs={null}
      type="string"
      typeRef={null}
    />
    <ModelComponentProperty
      key="organisasjonsform"
      name="organisasjonsform"
      oneOfTypeRefs={null}
      type={null}
      typeRef="#/definitions/Organisasjonsform"
    />
    <ModelComponentProperty
      key="postadresse"
      name="postadresse"
      oneOfTypeRefs={null}
      type={null}
      typeRef="#/definitions/Adresse"
    />
    <ModelComponentProperty
      key="registreringsdatoEnhetsregisteret"
      name="registreringsdatoEnhetsregisteret"
      oneOfTypeRefs={null}
      type="string"
      typeRef={null}
    />
    <ModelComponentProperty
      key="registrertIMvaregisteret"
      name="registrertIMvaregisteret"
      oneOfTypeRefs={null}
      type="boolean"
      typeRef={null}
    />
    <ModelComponentProperty
      key="naeringskode1"
      name="naeringskode1"
      oneOfTypeRefs={null}
      type={null}
      typeRef="#/definitions/Naeringskode"
    />
    <ModelComponentProperty
      key="naeringskode2"
      name="naeringskode2"
      oneOfTypeRefs={null}
      type={null}
      typeRef="#/definitions/Naeringskode"
    />
    <ModelComponentProperty
      key="naeringskode3"
      name="naeringskode3"
      oneOfTypeRefs={null}
      type={null}
      typeRef="#/definitions/Naeringskode"
    />
    <ModelComponentProperty
      key="antallAnsatte"
      name="antallAnsatte"
      oneOfTypeRefs={null}
      type="integer"
      typeRef={null}
    />
    <ModelComponentProperty
      key="forretningsadresse"
      name="forretningsadresse"
      oneOfTypeRefs={null}
      type={null}
      typeRef="#/definitions/Adresse"
    />
    <ModelComponentProperty
      key="stiftelsedato"
      name="stiftelsedato"
      oneOfTypeRefs={null}
      type="string"
      typeRef={null}
    />
    <ModelComponentProperty
      key="institusjonellSektorkode"
      name="institusjonellSektorkode"
      oneOfTypeRefs={null}
      type={null}
      typeRef="#/definitions/Sektorkode"
    />
    <ModelComponentProperty
      key="registrertIForetaksregisteret"
      name="registrertIForetaksregisteret"
      oneOfTypeRefs={null}
      type="boolean"
      typeRef={null}
    />
    <ModelComponentProperty
      key="registrertIStiftelsesregisteret"
      name="registrertIStiftelsesregisteret"
      oneOfTypeRefs={null}
      type="boolean"
      typeRef={null}
    />
    <ModelComponentProperty
      key="registrertIFrivillighetsregisteret"
      name="registrertIFrivillighetsregisteret"
      oneOfTypeRefs={null}
      type="boolean"
      typeRef={null}
    />
    <ModelComponentProperty
      key="sisteInnsendteAarsregnskap"
      name="sisteInnsendteAarsregnskap"
      oneOfTypeRefs={null}
      type="string"
      typeRef={null}
    />
    <ModelComponentProperty
      key="konkurs"
      name="konkurs"
      oneOfTypeRefs={null}
      type="boolean"
      typeRef={null}
    />
    <ModelComponentProperty
      key="underAvvikling"
      name="underAvvikling"
      oneOfTypeRefs={null}
      type="boolean"
      typeRef={null}
    />
    <ModelComponentProperty
      key="underTvangsavviklingEllerTvangsopplosning"
      name="underTvangsavviklingEllerTvangsopplosning"
      oneOfTypeRefs={null}
      type="boolean"
      typeRef={null}
    />
    <ModelComponentProperty
      key="maalform"
      name="maalform"
      oneOfTypeRefs={null}
      type="string"
      typeRef={null}
    />
    <ModelComponentProperty
      key="_links"
      name="_links"
      oneOfTypeRefs={null}
      type={null}
      typeRef="#/definitions/Self"
    />
  </Collapse>
</Fragment>
`;
